### PR TITLE
SwiftParser: A lookahead fix and test coverage for member types with non-identifier qualifiers

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Builds on #1077, adding and rearranging tests, and fixing an issue with `canParseType`.
Self-sufficient companion of apple/swift#62775

### CODE REVIEW

File: Tests/SwiftParserTest/translated/TypeExprTests.swift [hunk 1]
```diff
@@ -550,27 +579,117 @@ final class TypeExprTests: XCTestCase {
...
-        _ = (P1 & Int).self 
-        _ = (P1? & P2).self 
-        _ = (P1 & P2.Type).self 
-        _ = P1 & P2 -> P3
-        _ = P1 & P2 -> P3 & P1 -> Int
-        _ = (() -> P1 & P2).self // Ok
-        _ = (P1 & P2 -> P3 & P2).self 
-        _ = ((P1 & P2) -> (P3 & P2) -> P1 & Any).self // Ok
-      }
+      _ = P & Q
```
Comment (ahoppen) [inline Tests/SwiftParserTest/translated/TypeExprTests.swift:585]: Were all the test cases above moved/rewritten to some other file? I didn’t do a full search but I can’t find the equivalent of `(P1 & (P2)).self` straight away.

File: Sources/SwiftParser/Types.swift [hunk 2]
```diff
@@ -639,14 +643,88 @@ extension Parser {
...
+    if self.currentToken.isEllipsis {
+      self.consumeAnyToken()
+    }
+
+    return true
+  }
+
+  mutating func skipTypeAttributeList() {
+    var specifierProgress = LoopProgressCondition()
+    // TODO: Can we model isolated/_const so that they're specified in both canParse* and parse*?
```
Comment (ahoppen) [inline Sources/SwiftParser/Types.swift:659]: I hope to be working on exactly that in the next couple of days. 🤞

**Review by ahoppen — COMMENTED** [review]
Thanks for the PR and, in particular the firework of test cases 🎆

Changes are looking good to me, I’ve got a few stylistic comments inline and two questions out of curiosity:
- What was the test case that motivated this PR?
- I assume that you used the `parse` methods as blueprints for the `canParse` methods, right? I’m wondering what we can do to make sure these stay in-sync in the future.

File: Tests/SwiftParserTest/translated/TypeExprTests.swift (hunk 1 above)
Comment (AnthonyLatsis) [reply Tests/SwiftParserTest/translated/TypeExprTests.swift:585]: These are from the updates to the translated file in apple/swift#62775

File: Tests/SwiftParserTest/TypeMetatypeTests.swift [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  }
+
+  func testMisc() {
+    AssertParse(
+      """
+      arg.covariantAssocMetatype1 { (_: Any.Type.Type.Type) in }
+      """
+    )
+
+    AssertParse(
```
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply Tests/SwiftParserTest/TypeMetatypeTests.swift:196]: I see, sounds reasonable. Will do!

File: Tests/SwiftParserTest/ExpressionTypeTests.swift [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+final class ExpressionTypeTests: XCTestCase {
+  func testTypeExpression() {
+    AssertParse("_ = (any Sequence<Int>).self")
+  }
+
+  /// Test that `canParseType()` succeeds for these generic arguments.
+  // FIXME: This is suboptimal. Instead, we should verify that 'canParseType()'
+  // succeeds for all types in a valid parse. This is likely doomed to cause
+  // test failures with the current 'canParseType()' implementation and must be
+  // addressed separately.
```
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply Tests/SwiftParserTest/ExpressionTypeTests.swift:25]: Ah, I phrased this as if it was a PR comment. I tried implementing this verification before the rebase on top of #1077 and got some out-of-scope test failures related to attributes, hence the comment.

File: Sources/SwiftParser/Types.swift (hunk 2 above)
Comment (AnthonyLatsis) [reply Sources/SwiftParser/Types.swift:659]: Note that @bnbarham authored this TODO; I just hoisted the method closer to its new client.

File: Tests/SwiftParserTest/TypeMemberTests.swift [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+
+import SwiftSyntax
+import SwiftParser
+import XCTest
+
+final class MemberTypeTests: XCTestCase {
+  func testKeyword() {
+    AssertParse(
+      "MyType.class",
+      { TypeSyntax.parse(from: &$0) },
```
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply Tests/SwiftParserTest/TypeMemberTests.swift:21]: Yeah, it definitely could.

> Also: Should we decide to keep `AssertParseType`, I think it should be used here

Sure. I anticipated the discussion around `AssertParse*` and chose not to use them beyond the additions in this PR (this test case was simply moved).

**Comment by AnthonyLatsis** [comment]
> What was the test case that motivated this PR?

This PR is the companion of apple/swift#62775 and apple/swift#62319. All these test cases and the `canParseType` refactoring are leftovers from an implementation that I reverted after a fresh checkout made me realize that Ben had already done the job in #1077.

> I assume that you used the parse methods as blueprints for the canParse methods, right? I’m wondering what we can do to make sure these stay in-sync in the future.

I suppose `AssertParse` needs a new verification phase where it visits the all the valid type nodes of the tree and verifies that running `canParseType` on their source ranges succeeds.

File: Tests/SwiftParserTest/Assertions.swift [hunk 6]
```diff
@@ -335,6 +335,81 @@ func AssertParse(
   )
 }
 
+/// Same as `AssertParse` overload with a `(String) -> S` `parse`,
+/// parsing the resulting `String` as a `TypeSyntax`.
+func AssertParseType(
```
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply Tests/SwiftParserTest/Assertions.swift:340]: > If we decide to add these, where do we stop? Should we have `AssertParse` for every node? I could see an argument being made that we should have one for each syntax base kind (search for `SYNTAX_BASE_KIND`) but I’m pretty sure that we shouldn’t have one for attributes.

I think we should either add one for every base syntax or none at all, but I don’t like the implication of having to GYBify the file. I would argue that having a `parse` method on `AttributeSyntax` is useful: some attributes can accept non-trivial arguments which most definitely warrant focused tests.

[the remark this answers is not in the document]
Comment (bnbarham) [reply Tests/SwiftParserTest/Assertions.swift:340]: Sorry, I assume the ugly `{ *Syntax.parse(&$0) }` are from my test cases. Definitely should have just used `TypeSyntax.parse` 😅. Given that's possible, I agree with Alex - we should just have the one `AssertParse` and take `TypeSyntax.parse` as the argument.

As for what nodes should have `parse`... IMO it should only be top level nodes (decl, stmt, expr, pattern, type) with *maybe* a couple exceptions. Note that you don't have to match the whole syntax tree with the substructure match, so you could still have attribute tests with `@foo(...) let someVar` and just not match the `let someVar`. We could also introduce better tree matching to eg. allow wildcards.

Also regarding the non-trivial attribute arguments, IIUC we'll likely be changing attribute parsing quite soon to be as general as we can, rather have than specific attribute handling (not sure that will be possible for *all* attributes, but ideally it'll handle most of them).

**Comment by bnbarham** [comment]
> I assume that you used the parse methods as blueprints for the canParse methods, right? I’m wondering what we can do to make sure these stay in-sync in the future.

To be fair, this has been the case the entire time (and in my PR). I was also wondering if we could improve that situation, but I already went far enough down a rabbit hole in that PR so left it for later 😅. Ideally we wouldn't have the two separate implementations, but if there's no way around that we could at least verify them like @AnthonyLatsis suggested during testing. Not ideal, better than nothing though.

File: Tests/SwiftParserTest/Assertions.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply Tests/SwiftParserTest/Assertions.swift:340]: > Sorry, I assume the ugly `{ *Syntax.parse(&$0) }` are from my test cases. Definitely should have just used `TypeSyntax.parse` 😅. 

Never mind, that wasn’t really the trigger 🙆🏼‍♂️ — I was simply looking into reducing the line count per test.

> Given that's possible, I agree with Alex - we should just have the one `AssertParse` and take `TypeSyntax.parse` as the argument.

Deal 👍

> Note that you don't have to match the whole syntax tree with the substructure match, so you could still have attribute tests with `@foo(...) let someVar` and just not match the `let someVar`.

I know, but I’d rather have one `@foo(...) let someVar` test and ten `@foo(...)` tests if the *arguments* is the focus.

> We could also introduce better tree matching to eg. allow wildcards.

File: Tests/SwiftParserTest/translated/TypeExprTests.swift [hunk 7]
```diff
@@ -162,6 +180,17 @@ final class TypeExprTests: XCTestCase {
       }
       """
     )
+
+    AssertParse(
+      """
+      _ = (G<X>).Y.self
```
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply Tests/SwiftParserTest/translated/TypeExprTests.swift]: @ahoppen Are you fine with having multiple `AssertParse` calls per test function?

**Comment by AnthonyLatsis** [comment]
I happened to rethink my approach to testing while tending to the feedback. Let me know what you think!

File: Tests/SwiftParserTest/translated/TypeExprTests.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (ahoppen) [reply Tests/SwiftParserTest/translated/TypeExprTests.swift]: Yes, that’s totally fine.

**Review by ahoppen — APPROVED** [review]
Thanks for fixing this. It’s looking good to me 👍

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by AnthonyLatsis** [comment]
Forgot to lint the changes.. Should pass now.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Pull request merged**